### DISCUSSION TITLE
How to import capacitor plugin elegant in spa mode

### DISCUSSION THREAD

**Discussion by maocis** [quasarframework/quasar#13736] [CLI - Capacitor / Cordova modes]
Hello everyone
I am using Quasar CLI with vite
I want use camera in capacitor mode,so I import based on process.env like this:

```
if (process.env.MODE === 'capacitor') { 
    import('@capacitor/camera').then(async ({Camera, CameraResultType}) => {
        const image = await Camera.getPhoto({
            quality: 90,
            allowEditing: true,
            resultType: CameraResultType.Uri
        });
        console.log('image_src', image.webPath)
    })
```
### In capacitor mode:

1. In android enulator , the code work well, I can see the camera UI.
2. In browser, I get error below, althougth I had install  @ionic/pwa-elements and initialize it in boot file follow in documents:
`core-f86805ad.js:105 TypeError: Failed to fetch dynamically imported module:`

### In spa mode
I get error from vite:
`[plugin:vite:import-analysis] Failed to resolve import "@capacitor/camera" from "src\components\HistoryRecord.vue". Does the file exist?`

### My guess
I think in spa mode, the project import module from /quasar-project/node_modules, but capacitor plugin installed in /quasar-project/src-capacitor/node_modules,so I get error:`Failed to resolve import.` so I changed code like this:
```
if (process.env.MODE === 'capacitor') { 
    import('/src-capacitor/node_modules/@capacitor/camera').then(async ({Camera, CameraResultType}) => {
        const image = await Camera.getPhoto({
            quality: 90,
            allowEditing: true,
            resultType: CameraResultType.Uri
        });
        console.log('image_src', image.webPath)
    })
```
now, I get no error in spa mode, and the code worked well in capacitor mode too
**But I bet this solution is not the best solution, for it is so urgly**
**Please help me, How to handle this elegant? Thanks very much**

**Comment by metalsadman** [comment]
Lgtm.

**Reply by maocis** [reply]
please help me, I traped this several days!
I wrote a function in /src/boot/capacitor.js
```
export function getCapacitorModule(module_name) {
    console.log('getCapacitorModule')
    const files = import.meta.globEager(`/src-capacitor/node_modules/@capacitor/${module_name}/**/index.js`)
    const modules = {}
    for (const item of Object.values(files)) {
        for (const module in item) {
            modules[module] = item[module]
        }
    }
    return modules
}
```
when I used as below, I get error: SyntaxError: Unexpected identifier (at
```
import getCapacitorModule from 'boot/capacitor' 
const modules = capacitor.getCapacitorModule('camera')
const image = await modules.Camera.getPhoto({
      quality: 90,
      allowEditing: true,
      resultType: modules.CameraResultType.Uri
});
```
then I changed import with require：
`const {capacitor} = require("boot/capacitor");`
But I get error again: Uncaught (in promise) ReferenceError: require is not defined
I am lost, how to handle this after all
thanks a lot!

**Comment by yusufkandemir** [comment]
Vite is probably only checking for `import.meta.env.*` usage for import analysis. So, it doesn't understand if the related code will be stripped out or not when `process.env.MODE === 'capacitor'`.

I would recommend trying the following to see if it makes a difference:
```js
// quasar.config.js > build
extendViteConf (config) {
  config.define['import.meta.env.MODE'] = config.define['process.env.MODE']
}
```

```js
// Your .vue, .js, etc. file
if (import.meta.env.MODE === 'capacitor') { 
  import('@capacitor/camera').then(async ({ Camera, CameraResultType }) => {
    const image = await Camera.getPhoto({
      quality: 90,
      allowEditing: true,
      resultType: CameraResultType.Uri
    });
    console.log('image_src', image.webPath)
  });
}
```
If this one works, you can safely use this approach while we are trying to find a better solution.

I think Vite should respect anything with `cfg.define`, not just `import.meta.env` or `process.env`, so we might want to check out for existing feature requests and open one if there is none. We can try to work around this in the meanwhile. If the Vite team doesn't accept this feature request, or it takes too long to implement, we can try to work around it by making some updates to app-vite.

If it doesn't work, Vite probably doesn't understand `SOME_ENV_VAR === 'comparison'` stuff, which might need a bug report in the Vite repo.
